internal/pkg/manifest/union.go

```diff
@@ -62,48 +66,64 @@ func (t Union[_, _]) IsAdvanced() bool {
 }
 
 // UnmarshalYAML decodes value into either type Basic or Advanced, and stores that value
-// in t. value is first decoded into type Basic, and t will hold type Basic if:
-//   - There was no error decoding value into type Basic
-//   - Basic.IsZero() returns false OR Basic does not support IsZero()
+// in t. value is first decoded into type Basic, and t will hold type Basic if
+// (1) There was no error decoding value into type Basic and
+// (2) Basic.IsZero() returns false OR Basic is not zero via reflection.
 //
 // If Basic didn't meet the above criteria, then value is decoded into type Advanced.
 // t will hold type Advanced if Advanced meets the same conditions that were required for type Basic.
 //
 // If value fails to decode into either type, or both types are zero after
 // decoding, t will not hold any value.
 func (t *Union[Basic, Advanced]) UnmarshalYAML(value *yaml.Node) error {
-	// reset struct
-	var simple Basic
-	var advanced Advanced
-	t.isBasic, t.Basic = false, simple
-	t.isAdvanced, t.Advanced = false, advanced
+	// convert value to []byte so we can use KnownFields()
+	// see https://github.com/go-yaml/yaml/issues/460
+	// side effect: TypeError line numbers are reset.
+	b, err := yaml.Marshal(value)
+	if err != nil {
+		return err
+	}
 
-	sErr := value.Decode(&simple)
-	if sErr == nil && !isZeroerAndZero(simple) {
-		t.isBasic, t.Basic = true, simple
+	var basic Basic
+	dec := yaml.NewDecoder(bytes.NewReader(b))
+	dec.KnownFields(true)
+	bErr := dec.Decode(&basic)
```

Comment:
Why do we do this 🤔 What will happen if we don't use `KnownFields`?

Reply:
ideally, it's something you set at the top level and the setting propagates down to custom `UnmarshalYAML()` functions. However, due to https://github.com/go-yaml/yaml/issues/460, the `value` passed in to `UnmarshalYAML()` doesn't retain the setting.

If we don't set `KnownFields()`, and we were unmarshalling
```yaml
key:
  invalid_key: asdf
```
into `Union[string, aStruct]`, then the value of Union would be aStruct, even though it's zero.

---

writing this, i realized that this statement:
> then the value of Union would be aStruct, even though it's zero.

is no longer true, because we are now using `reflect.IsZero()` when checking if a value is zero, which would be true if the struct is full of zero value fields. i guess using `KnownFields()` is just a semantic change in that case - erroring if an invalid field is set.